### PR TITLE
Only uncover password on focus

### PR DESCRIPTION
**Pull request by christopheschwyzer**
Follow up https://github.com/cargomedia/cm/pull/1854

To make password managers (like 1Password) work.

### CODE REVIEW

**Comment by christopheschwyzer** [comment]
@njam @vogdb maybe there is a better way to do this?!

**Comment by njam** [comment]
- Let's not pass an `event` instance from the event handler to the function (as it makes the function unusable from other places where we don't have an event).
 - Rather I would put *some* logic into the `focus` event handler, à la `if (length === 0) { togglePasswordMask(); }`
- Don't we have the problem now that on every `focus` event we toggle the mode if there is some text?

**Comment by christopheschwyzer** [comment]
done w/ @njam 
@vogdb wanna have a look?

**Comment by njam** [comment]
@christopheschwyzer I think it would be good to hide the password on "paste" as you suggested.
-> https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/onpaste

**Comment by vogdb** [comment]
looks ok.

**Comment by christopheschwyzer** [comment]
I added 
```
      'paste input': function() {
            this.togglePasswordVisibility(false);
        }
```
Now when I paste some text from clipboard it works fine - when I use 1PW it doesn't switch to password. Any ideas?

**Comment by vogdb** [comment]
It can actually type the password, not paste it =) It just does it real quick cause it's a program.

**Comment by njam** [comment]
Oh, then that's not an option  :/
Either we ignore it, or we detect input speed?

**Comment by vogdb** [comment]
For that we need to be sure that 1PW types it and not sets it directly smth like this: `input.value = value`. Anyway we don't want to depend on its inner workings. Detection of speed is not an option.

**Comment by njam** [comment]
That would trigger our code as well (because we're using `propertychange` in `CM_FormField_Text.enableTriggerChangeOnInput()`).
We could detect that the text changes from 0 to 2 characters in less than X milliseconds, then switch to password mode?

**Comment by vogdb** [comment]
Is it ok that now browsers' password managers do not offer to save password on our Login and SignUp forms?

**Comment by njam** [comment]
Yes, I noticed that too :(
But I think with this pull request it should be fixed, as we're using `type="password"` again by default, can you confirm?

**Comment by vogdb** [comment]
hm, didn't work for me. I don't see why it is suppose to work. When you start to type, input becomes of `type = "text"` and does not return back to `password` before submitting.

**Comment by vogdb** [comment]
btw why we have a new formatting here? 4 spaces instead of 2? Did we change our default formatting?

**Comment by vogdb** [comment]
I pushed here a small change to enable Password Manager. Verified it on the latest Chrome. Works.

**Comment by christopheschwyzer** [comment]
So you change back to password on submit? Good idea!

**Comment by vogdb** [comment]
Found an interesting bug on Android. The first entered symbol goes forward on the right side of the cursor. It's like this symbol was already in the input field before we had started to type.

Also have a question. Do we need to change input type to `text` in case of submit error? Now we change it to `password` before submit.

**Comment by vogdb** [comment]
@christopheschwyzer @njam  please look at the change.

**Comment by christopheschwyzer** [comment]
> Also have a question. Do we need to change input type to text in case of submit error? Now we change it to password before submit.

This will not be possible b/c error is only possible after submit?!

**Comment by vogdb** [comment]
My last commit stores the status of `visibilityDesired` on `submit` and restores it on `error`.

**Comment by christopheschwyzer** [comment]
Ah I see, smart. Didn't notice your commit. Sounds reasonable.

**Comment by njam** [comment]
@tomaszdurka could you review this? Aleksei, maybe give Tomasz an intro on the topic?

**Comment by tomaszdurka** [comment]
Isn't Aleksei on holidays this week? I am aware of the issue. 

*I am generally against uncovering password, but well, you guys probably got this topic covered right* :panda_face: Shouldn't this be only for registration?

File: library/CM/FormField/Password.js [hunk 1]
```diff
@@ -5,18 +5,62 @@
...
+      this.togglePasswordVisibility(false);
+    }
+    CM_FormField_Text.prototype.triggerChange.apply(this, arguments);
+  },
+
+  /**
+   * @param {Boolean} [state]
+   */
+  togglePasswordVisibility: function(state) {
+    if ('undefined' === typeof state) {
```
Comment (tomaszdurka) [inline library/CM/FormField/Password.js]: There is underscore function for this too: `_.isUndefined()`

File: library/CM/FormField/Password.js [hunk 2]
```diff
@@ -5,18 +5,62 @@
...
+  },
 
-    var showText = ($input.attr('type') === 'password');
-    $input.attr('type', showText ? 'text' : 'password');
-    $buttonText.toggle(showText);
-    $buttonPassword.toggle(!showText);
+  /**
+   * @param {Boolean} state
+   */
+  _togglePasswordVisibilityInput: function(state) {
```
Comment (tomaszdurka) [inline library/CM/FormField/Password.js]: Maybe call this `_setInputTypeByState`?

**Comment by vogdb** [comment]
> Shouldn't this be only for registration?

It's hard for me to answer this question. I guess guys thought about it.

**Comment by vogdb** [comment]
@tomaszdurka please review

**Comment by christopheschwyzer** [comment]
@tomaszdurka Can you elaborate why you think it's not a good idea?

**Comment by vogdb** [comment]
@christopheschwyzer lets finish this issue without Tomasz. It might take a while otherwise.

**Comment by tomaszdurka** [comment]
@vogdb Let's say it's good.

@christopheschwyzer I could ask why you guys actually introduced this feature in the first place :smile_cat: Is this a trend on other platforms? 
Personally I feel that password is something very private (even for the non-important service you signed up for fun) and exposing it to anyone/anything except your mind feels unsecure. Especially in times when privacy is so important and it's lack might feel unsafe and dangerous. When I see password revealed - my mind somehow translates it to plain text passwords stored with no real security. 
Is this really only me thinking this way? :panda_face:

**Comment by christopheschwyzer** [comment]
@tomaszdurka I agree on the first feeling that plain text is insecure. But then you still have the option to hide it if you're uncomfortable. It's a very rare situation that you have somebody looking over your shoulder while you type in your password. From UX perspective I'd say you help a lot of people from misspelling the pw, especially on signup, where it is crucial. About the _trend_ - I see more and more  "Reveal password" buttons, and I think it's only the first step. You can think of us as trendsetters in that regard. ;)

**Comment by vogdb** [comment]
@christopheschwyzer Are you ok with the changes? Mergeable?

**Comment by christopheschwyzer** [comment]
Doing some final testing right now.

**Comment by christopheschwyzer** [comment]
@vogdb Works fine to input the pw. I noticed that 1PW doesn't ask you to store the pw the first time you input it (b/c it's not a pw input). Can we change it to be a pw type before/while we submit?

**Comment by vogdb** [comment]
hm, we do it before submit in this PR.

**Comment by christopheschwyzer** [comment]
True. So we change it too late somehow? Is there another event 1PW might use?

**Comment by tomaszdurka** [comment]
Let's cover by default - solves all the problems :panda_face:

**Comment by vogdb** [comment]
> So we change it too late somehow?

That seems to be the truth. If you hide password field then 1PW works fine. It does not matter when you do it. After or before password is filled. The only thing that matters is that it has `password` type before submit =(.

**Comment by vogdb** [comment]
Changing it to `password` by default will not fix the problem completely cause when user changes the type to `text` before submit 1PW won't work either.

**Comment by NicolasSchmutz** [comment]
TBH I would not bother too much about 1PW (although I love it too). 
I don't think our users use such a tool. I see more of an improvement for mobile users if it is not covered by default...

**Comment by vogdb** [comment]
We can divide this behavior and make it uncover for small displays and cover for big displays. Nevertheless the 1PW problem exists.

**Comment by vogdb** [comment]
Hm, now Chrome password manager does not work anymore. Firefox still works. I'm afraid that this feature makes unstable usage of password managers. And I think no one will argue that a big portion of users relies on password managers.

**Comment by vogdb** [comment]
If password field was rendered from the beginning as the password and we don't change it's type in `ready` of CM_FormField_Password then browsers managers work. 1PW still does not. So, I'll propose to change the default visibility to hidden cause it fixes password managers problems.

**Comment by vogdb** [comment]
@NicolasSchmutz @christopheschwyzer I need your opinions here.

**Comment by christopheschwyzer** [comment]
I see the benefits don't make up for the caused problems :/ Let's switch the behavior after all. Password by default. Possibility to uncover. Switch to PW on submit and ignoring that it won't work 100% on all Browsers. @tomaszdurka @vogdb @NicolasSchmutz agree?

**Comment by NicolasSchmutz** [comment]
sgmt

**Comment by vogdb** [comment]
@christopheschwyzer please review.

**Comment by christopheschwyzer** [comment]
lgtm. @tomaszdurka please verify

**Comment by vogdb** [comment]
@tomaszdurka please purify

**Comment by tomaszdurka** [comment]
Sounds looks good.

**Comment by christopheschwyzer** [comment]
Small details. Maybe we should re-focus the field after changing the type. @vogdb wdyt?

**Comment by vogdb** [comment]
yep, that will be good

**Comment by vogdb** [comment]
@christopheschwyzer please review.

**Comment by christopheschwyzer** [comment]
lgtm!

**Pull request merged**